File: _includes/code/python/howto.configure.rbac.roles.py
from weaviate.classes.rbac import Permissions

# START AdminClient
import weaviate
from weaviate.classes.init import Auth

# Connect to Weaviate as root user
client = weaviate.connect_to_local(
    # END AdminClient
    # Use custom port defined in tests/docker-compose-rbac.yml (without showing the user)
    port=8580,
    grpc_port=50551,
    # START AdminClient
    auth_credentials=Auth.api_key("user-a-key"),
)
# END AdminClient

# TODO: Remove if not used
custom_user_client = weaviate.connect_to_local(
    port=8580, grpc_port=50551, auth_credentials=Auth.api_key("user-b-key")
)

all_roles = client.roles.list_all()
for role_name, _ in all_roles.items():
    if role_name not in ["viewer", "root", "admin"]:
        client.roles.delete(role_name=role_name)

# # START CreateRole
# client.roles.create(role_name="testRole")
# # END CreateRole

# START AddManageRolesPermission
from weaviate.classes.rbac import Permissions, RoleScope

permissions = [
    Permissions.roles(
        role="testRole*",  # Applies to all roles starting with "testRole"
        create=True,  # Allow creating roles
        read=True,  # Allow reading roles
        update=True,  # Allow updating roles
        delete=True,  # Allow deleting roles
        scope=RoleScope.MATCH,  # Only allow role management with the current user's permission level
        # scope=RoleScope.ALL   # Allow role management with all permissions
    )
]

client.roles.create(role_name="testRole", permissions=permissions)
# END AddManageRolesPermission

assert "testRole" in client.roles.list_all().keys()

client.roles.delete("testRole")

# START AddCollectionsPermission
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.collections(
        collection="TargetCollection*",  # Applies to all collections starting with "TargetCollection"
        create_collection=True,  # Allow creating new collections
        read_config=True,  # Allow reading collection info/metadata
        update_config=True,  # Allow updating collection configuration, i.e. update schema properties, when inserting data with new properties
        delete_collection=True,  # Allow deleting collections
    ),
]

client.roles.create(role_name="testRole", permissions=permissions)
# END AddCollectionsPermission
permissions = client.roles.get(role_name="testRole")
assert any(
    permission.collection == "TargetCollection*"
    for permission in permissions.collections_permissions
)

client.roles.delete("testRole")

# START AddTenantPermission
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.collections(
        collection="TargetCollection*",
        create_collection=True,
        read_config=True,
        update_config=True,
        delete_collection=True,
    ),
    Permissions.tenants(
        collection="TargetCollection*",  # Applies to all collections starting with "TargetCollection"
        create=True,  # Allow creating new tenants
        read=True,  # Allow reading tenant info/metadata
        update=True,  # Allow updating tenant states
        delete=True,  # Allow deleting tenants
    ),
]

client.roles.create(role_name="testRole", permissions=permissions)
# END AddTenantPermission
permissions = client.roles.get(role_name="testRole")
assert any(
    permission.collection == "TargetCollection*"
    for permission in permissions.collections_permissions
)

client.roles.delete("testRole")

# START AddDataObjectPermission
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.data(
        collection="TargetCollection*",  # Applies to all collections starting with "TargetCollection"
        create=True,  # Allow data inserts
        read=True,  # Allow query and fetch operations
        update=True,  # Allow data updates
        delete=False,  # Allow data deletes
    ),
]

client.roles.create(role_name="testRole", permissions=permissions)
# END AddDataObjectPermission
permissions = client.roles.get(role_name="testRole")
assert any(
    permission.collection == "TargetCollection*"
    for permission in permissions.data_permissions
)

client.roles.delete("testRole")

# START AddBackupPermission
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.backup(
        collection="TargetCollection*",  # Applies to all collections starting with "TargetCollection"
        manage=True,  # Allow managing backups
    ),
]

client.roles.create(role_name="testRole", permissions=permissions)
# END AddBackupPermission
permissions = client.roles.get(role_name="testRole")
assert any(
    permission.collection == "TargetCollection*"
    for permission in permissions.backups_permissions
)

client.roles.delete("testRole")

# START AddClusterPermission
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.cluster(read=True),  # Allow reading cluster data
]

client.roles.create(role_name="testRole", permissions=permissions)
# END AddClusterPermission
permissions = client.roles.get(role_name="testRole")
assert permissions.cluster_permissions

client.roles.delete("testRole")

# START AddNodesPermission
from weaviate.classes.rbac import Permissions

verbose_permissions = [
    Permissions.Nodes.verbose(
        collection="TargetCollection*",  # Applies to all collections starting with "TargetCollection"
        read=True,  # Allow reading node metadata
    ),
]

# The `minimal` verbosity level applies to all collections unlike
# the `verbose` level where you specify the collection name filter
minimal_permissions = [
    Permissions.Nodes.minimal(
        read=True,  # Allow reading node metadata
    ),
]

client.roles.create(role_name="testRole", permissions=verbose_permissions)  # or `minimal_permissions`
# END AddNodesPermission

permissions = client.roles.get(role_name="testRole")
print("ivan ", permissions)
assert any(
    permission.collection == "TargetCollection*"
    for permission in permissions.nodes_permissions
)


client.roles.delete("testRole")

permissions = [
    Permissions.collections(
        collection="TargetCollection*",
        read_config=True,
    ),
]

client.roles.create(role_name="testRole", permissions=permissions)

# START AddRoles
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.data(collection="TargetCollection*", read=True, create=True),
    Permissions.data(collection="TargetCollection*", read=True, create=False),
]

client.roles.add_permissions(permissions=permissions, role_name="testRole")
# END AddRoles

# START AssignRole
client.users.assign_roles(user_id="user-b", role_names=["testRole", "viewer"])
# END AssignRole
assert "testRole" in client.users.get_assigned_roles("user-b")
assert "viewer" in client.users.get_assigned_roles("user-b")

# START ListCurrentUserRoles
print(client.users.get_my_user())
# END ListCurrentUserRoles

# START ListUserRoles
user_roles = client.users.get_assigned_roles("user-b")

for role in user_roles:
    print(role)
# END ListUserRoles
assert any(
    permission.collection == "TargetCollection*"
    for permission in user_roles["testRole"].collections_permissions
)
assert any(
    permission.collection == "TargetCollection*"
    for permission in user_roles["testRole"].data_permissions
)

# START CheckRoleExists
print(client.roles.exists(role_name="testRole"))  # Returns True or False
# END CheckRoleExists

# START InspectRole
test_role = client.roles.get(role_name="testRole")

print(test_role)
print(test_role.collections_permissions)
print(test_role.data_permissions)
# END InspectRole

# START AssignedUsers
assigned_users = client.roles.get_assigned_user_ids(role_name="testRole")

for user in assigned_users:
    print(user)
# END AssignedUsers
assert "user-b" in assigned_users

# START ListAllRoles
all_roles = client.roles.list_all()

for role_name, role in all_roles.items():
    print(role_name, role)
# END ListAllRoles

# START RemovePermissions
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.collections(
        collection="TargetCollection*",
        read_config=True,
        create_collection=True,
        delete_collection=True,
    ),
    Permissions.data(collection="TargetCollection*", read=True, create=False),
]

client.roles.remove_permissions(role_name="testRole", permissions=permissions)
# END RemovePermissions

# START RevokeRoles
client.users.revoke_roles(user_id="user-b", role_names="testRole")
# END RevokeRoles
assert "testRole" not in client.users.get_assigned_roles("user-b")

# START DeleteRole
client.roles.delete(role_name="testRole")
# END DeleteRole

client.close()
custom_user_client.close()

File: _includes/code/typescript/howto.configure.rbac.roles.ts
import assert from "assert"
// START AdminClient
import weaviate, { WeaviateClient } from 'weaviate-client'

// Connect to Weaviate as root user
const client: WeaviateClient = await weaviate.connectToLocal({
    // END AdminClient
    // Use custom port defined in tests/docker-compose-rbac.yml (without showing the user)
    port: 8580,
    grpcPort: 50551,
    // START AdminClient
    authCredentials: new weaviate.ApiKey("user-a-key")
})
// END AdminClient

// TODO: Remove if not used
const customUserClient: WeaviateClient = await weaviate.connectToLocal({
    port: 8580,
    grpcPort: 50551,
    authCredentials: new weaviate.ApiKey("user-b-key")
})

const allRolesCheck = await client.roles.listAll()

for await (const [key, value] of Object.entries(allRolesCheck)) {
    if (!["viewer", "root", "admin"].includes(key)) {
        await client.roles.delete(key)
    }
}

// Todo: This will be added in upcoming release
// START CreateRole
await client.roles.create("testRole")
// END CreateRole

// START AddClusterPermission // START AddManageRolesPermission // START AddCollectionsPermission // START AddTenantPermission // START AddDataObjectPermission // START AddBackupPermission // START AddNodesPermission // START AddRoles // START RemovePermissions
const { permissions } = weaviate
// END AddClusterPermission // END AddManageRolesPermission // END AddCollectionsPermission // END AddTenantPermission // END AddDataObjectPermission // END AddBackupPermission // END AddNodesPermission // END AddRoles // END RemovePermissions

// todo add scope when tommy adds it
// START AddManageRolesPermission

const rolePermission = permissions.roles({
    role: "testRole",
    create: true,
    read: true,
    update: true,
    delete: true,
})

await client.roles.create("testRole", rolePermission)
// END AddManageRolesPermission

assert(Object.keys(await client.roles.listAll()).includes('testRole'));

await client.roles.delete("testRole")

// START AddCollectionsPermission

const collectionPermissions = [
    permissions.collections({
        collection: "TargetCollection*",  // Applies to all collections starting with "TargetCollection"
        create_collection: true,  // Allow creating new collections
        read_config: true,  // Allow reading collection info/metadata
        update_config: true,  // Allow updating collection configuration, i.e. update schema properties, when inserting data with new properties
        delete_collection: true,  // Allow deleting collections
    }),
]

await client.roles.create("testRole", collectionPermissions)

// END AddCollectionsPermission
const getCollectionPermissions = await client.roles.byName("testRole")

if (getCollectionPermissions) {
    assert.equal((getCollectionPermissions.dataPermissions.some(
        permission => permission.collection == "TargetCollection*"
    )), true)
}

await client.roles.delete("testRole")

// START AddTenantPermission

const AddTenantPermissions = [
    permissions.tenants({
        collection: "TargetCollection*",  // Applies to all collections starting with "TargetCollection"
        tenant: "TargetTenant*",  // Applies to all tenants starting with "TargetTenant"
        create: true,  // Allow creating new tenants
        read: true,  // Allow reading tenant info/metadata
        update: true,  // Allow updating tenant states
        delete: true,  // Allow deleting tenants
    }),
]

await client.roles.create("testRole", AddTenantPermissions)
// END AddTenantPermission
const getTenantCollection = await client.roles.byName("testRole")

if (getTenantCollection) {
    assert.equal((getTenantCollection.dataPermissions.some(
        permission => permission.collection == "TargetCollection*"
    )), true)
}

client.roles.delete("testRole")

// START AddDataObjectPermission

const dataPermissions = [
    permissions.data({
        collection: "TargetCollection*",  // Applies to all collections starting with "TargetCollection"
        tenant: "TargetTenant*",  // Applies to all tenants starting with "TargetTenant"
        create: true,  // Allow data inserts
        read: true,  // Allow query and fetch operations
        update: true,  // Allow data updates
        delete: false,  // Allow data deletes
    }),
]

await client.roles.create("testRole", dataPermissions)

// END AddDataObjectPermission
const getDataPermissions = await client.roles.byName("testRole")

if (getDataPermissions) {
    assert.equal((getDataPermissions.dataPermissions.some(
        permission => permission.collection == "TargetCollection*"
    )), true)
}

await client.roles.delete("testRole")

// START AddBackupPermission

const backupsPermissions = [
    permissions.backup({
        collection: "TargetCollection*",  // Applies to all collections starting with "TargetCollection"
        manage: true,  // Allow managing backups
    }),
]

await client.roles.create("testRole", backupsPermissions)
// END AddBackupPermission

const getBackupsPermissions = await client.roles.byName("testRole")

if (getBackupsPermissions) {
    assert.equal((getBackupsPermissions.dataPermissions.some(
        permission => permission.collection == "TargetCollection*"
    )), true)
}

await client.roles.delete("testRole")

// START AddClusterPermission

const clusterPermissions = [
    permissions.cluster({
        read: true // Allow reading cluster data
    }),
]

await client.roles.create("testRole", clusterPermissions)
// END AddClusterPermission

const getClusterPermissions = await client.roles.byName("testRole")
// assert permissions.cluster_permissions

await client.roles.delete("testRole")

// START AddNodesPermission

const verboseNodePermissions = [
    permissions.nodes.verbose({
        collection: "TargetCollection*",  // Applies to all collections starting with "TargetCollection"
        read: true,  // Allow reading node metadata
    }),
]

// The `minimal` verbosity level applies to all collections unlike
// the `verbose` level where you specify the collection name filter
const minimalNodePermissions = [
    permissions.nodes.minimal({
        read: true,  // Allow reading node metadata
    }),
]

await client.roles.create("testRole", verboseNodePermissions)  // or `minimalNodePermissions`
// END AddNodesPermission

const getNodePermissions = await client.roles.byName("testRole")

if (getNodePermissions) {
    assert.equal((getNodePermissions.dataPermissions.some(
        permission => permission.collection == "TargetCollection*"
    )), true)
}


await client.roles.delete("testRole")

// This is to add additional permission to below
const dummyPermission = [
    permissions.collections({
        collection: "TargetCollection*",
        read_config: true,
    }),
]

await client.roles.create("testRole", dummyPermission)

// START AddRoles

const additionalDataPermissions = [
    permissions.data({
        collection: "TargetCollection*",
        read: true,
        create: true
    }),
    permissions.data({
        collection: "TargetCollection*",
        read: true,
        create: false
    }),
]

client.roles.addPermissions("testRole", additionalDataPermissions)
// END AddRoles

// START AssignRole
await client.users.assignRoles(["testRole", "viewer"], "user-b")
// END AssignRole
assert.equal((Object.keys(await client.users.getAssignedRoles("user-b")).some(
    role => role == "viewer*"
)), true)

assert.equal((Object.keys(await client.users.getAssignedRoles("user-b")).some(
    role => role == "testRole"
)), true)

// START ListCurrentUserRoles
console.log(await client.users.getMyUser())
// END ListCurrentUserRoles

// START ListUserRoles
const userRoles = await client.users.getAssignedRoles("user-b")

for (const [role, value] of Object.entries(userRoles)) {
    console.log(role)
}
// END ListUserRoles

assert.equal((userRoles["testRole"].collectionsPermissions.some(
    permission => permission.collection == "TargetCollection*"
)), true)

assert.equal((userRoles["testRole"].dataPermissions.some(
    permission => permission.collection == "TargetCollection*"
)), true)

// START CheckRoleExists
console.log(await client.roles.exists("testRole"))  // Returns true or false
// END CheckRoleExists

// START InspectRole 
const testRole = await client.roles.byName("testRole")

console.log(testRole)
console.log(testRole?.collectionsPermissions)
console.log(testRole?.dataPermissions)
// END InspectRole

// START AssignedUsers
const assignedUsers = await client.roles.assignedUserIds("testRole")

for (const users of assignedUsers) {
    console.log(users)
}
// END AssignedUsers
assert.equal(assignedUsers.some(
    role => role == "viewer*"
), true)

// START ListAllRoles
const allRoles = await client.roles.listAll()

for (const [key, value] of Object.entries(allRoles)) {
    console.log(key)
}
// END ListAllRoles

// START RemovePermissions

const permissionsToRemove = [
    permissions.collections({
        collection: "TargetCollection*",
        read_config: true,
        create_collection: true,
        delete_collection: true,
    }),
    permissions.data({
        collection: "TargetCollection*",
        read: true,
        create: false
    }),
]

await client.roles.removePermissions("testRole", permissionsToRemove)
// END RemovePermissions

// START RevokeRoles
await client.users.revokeRoles("user-b", "testRole")
// END RevokeRoles
assert.equal((Object.keys(await client.users.getAssignedRoles("user-b")).some(
    role => role == "testRole"
)), false)

// START DeleteRole
await client.roles.delete("testRole")
// END DeleteRole

client.close()
customUserClient.close()
